File: apps/web/src/types/meeting_brief.ts
// Mirrors apps/api/app/schemas/meeting_brief.py.

export type BriefCallType =
  | "first_discovery"
  | "qbr"
  | "renewal"
  | "expansion"
  | "other";

export type Severity = "high" | "caution";
export type AttendeeCompany = "client" | "beroe";
export type ScenarioType = "good" | "neutral" | "poor";

export interface SnapshotStat {
  num: string;
  label: string;
  sub: string | null;
}

export interface CallTimerSlot {
  time: string;
  label: string;
}

export interface Attendee {
  initials: string;
  name: string;
  role: string | null;
  company: AttendeeCompany;
  is_self: boolean;
  avatar_color: string | null;
  objectives: string[];
  primary_objective: string | null;
  background: string[];
  opening_ask: string | null;
}

export interface Minefield {
  severity: Severity;
  type: string | null;
  text: string;
  why: string | null;
}

export interface Objective {
  rank: number;
  name: string;
  confidence: number; // 1..5
  bullets: string[];
  beroe: string | null;
  sources: string[];
}

export interface DiscoveryQuestion {
  objective: string;
  rank: number;
  person: string;
  from_email: boolean;
  text: string;
  // H46 — Category dropdown on Discovery questions.
  category?: string | null;
}

export const DISCOVERY_CATEGORIES = [
  "Commercial",
  "Risk",
  "People",
  "Process",
  "Sustainability",
  "Technology",
  "Adoption",
] as const;

export type DiscoveryCategory = (typeof DISCOVERY_CATEGORIES)[number];

export interface ValueAnchorPoint {
  text: string;
  note: string | null;
}

export interface ValueAnchor {
  objective: string;
  points: ValueAnchorPoint[];
}

export interface EmailInsight {
  meta: string;
  bullets: string[];
}

export interface PublicSignal {
  person: string | null;
  headline: string;
  text: string | null;
  url: string | null;
  tag: string | null;
}

export interface NewsItem {
  days_ago: number | null;
  headline: string;
  source: string | null;
  signal: string | null;
  url: string | null;
  tag: string | null;
}

export interface AnnualReportItem {
  title: string;
  year: number | null;
  url: string | null;
  bullets: string[];
}

export interface ClosingScenario {
  type: ScenarioType;
  label: string | null;
  text: string;
}

export interface MeetingBrief {
  account_id: string;

  call_type: BriefCallType | null;
  call_date: string | null; // ISO yyyy-mm-dd
  call_time: string | null;
  call_platform: string | null;
  call_duration_minutes: number | null;

  win_condition: string | null;
  cheat_sheet_win_condition_short: string | null;

  company_snapshot: SnapshotStat[];
  call_timer: CallTimerSlot[];
  attendees: Attendee[];
  minefields: Minefield[];
  objectives: Objective[];
  discovery_questions: DiscoveryQuestion[];
  value_anchors: ValueAnchor[];
  email_insights: EmailInsight[];
  public_signals: PublicSignal[];
  news: NewsItem[];
  annual_reports: AnnualReportItem[];
  closing_scenarios: ClosingScenario[];
  cheat_sheet_never_say: string[];
  cheat_sheet_opening_asks: string[];
  // H46 — Categories tab.
  categories: string[];

  updated_at: string;
  updated_by: string | null;
  is_editable: boolean;
}

export type MeetingBriefUpdate = Partial<
  Omit<MeetingBrief, "account_id" | "updated_at" | "updated_by" | "is_editable">
>;

export const BRIEF_CALL_TYPE_LABELS: Record<BriefCallType, string> = {
  first_discovery: "First Discovery",
  qbr: "QBR",
  renewal: "Renewal",
  expansion: "Expansion",
  other: "Other",
};

export const SEVERITY_LABELS: Record<Severity, string> = {
  high: "High",
  caution: "Caution",
};

export const SCENARIO_LABELS: Record<ScenarioType, string> = {
  good: "Strong close",
  neutral: "Warm handoff",
  poor: "Keep door open",
};

/** Empty brief used when the server returns null collections on a fresh row. */
export function emptyBrief(accountId: string): MeetingBrief {
  return {
    account_id: accountId,
    call_type: null,
    call_date: null,
    call_time: null,
    call_platform: null,
    call_duration_minutes: null,
    win_condition: null,
    cheat_sheet_win_condition_short: null,
    company_snapshot: [],
    call_timer: [],
    attendees: [],
    minefields: [],
    objectives: [],
    discovery_questions: [],
    value_anchors: [],
    email_insights: [],
    public_signals: [],
    news: [],
    annual_reports: [],
    closing_scenarios: [],
    cheat_sheet_never_say: [],
    cheat_sheet_opening_asks: [],
    categories: [],
    updated_at: new Date().toISOString(),
    updated_by: null,
    is_editable: false,
  };
}

File: apps/api/app/schemas/meeting_brief.py
"""M12 — Pre-Meeting Brief schemas. Validates the JSONB collection shapes.

Each nested model mirrors one row inside its JSONB column. Pydantic v2
runs structural validation on PATCH; bad payloads return 422 without
ever hitting the DB.

Naming: we use snake_case on the wire so the JSONB stays Pythonic. The
v20 prototype used camelCase (callType, winCondition, primaryObjective);
we translate that on the frontend.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

BriefCallType = Literal["first_discovery", "qbr", "renewal", "expansion", "other"]
Severity = Literal["high", "caution"]
AttendeeCompany = Literal["client", "beroe"]
ScenarioType = Literal["good", "neutral", "poor"]


# ---------- Nested item shapes (one per JSONB column) ----------


class SnapshotStat(BaseModel):
    num: str = Field(..., max_length=40)
    label: str = Field(..., max_length=120)
    sub: str | None = Field(None, max_length=120)


class CallTimerSlot(BaseModel):
    time: str = Field(..., max_length=40)
    label: str = Field(..., max_length=200)


class Attendee(BaseModel):
    initials: str = Field(..., max_length=4)
    name: str = Field(..., max_length=120)
    role: str | None = Field(None, max_length=160)
    company: AttendeeCompany
    is_self: bool = False
    avatar_color: str | None = Field(None, max_length=24)  # hex or token
    objectives: list[str] = Field(default_factory=list)
    primary_objective: str | None = Field(None, max_length=80)
    background: list[str] = Field(default_factory=list)
    opening_ask: str | None = Field(None, max_length=600)


class Minefield(BaseModel):
    severity: Severity
    type: str | None = Field(None, max_length=60)
    text: str = Field(..., max_length=400)
    why: str | None = Field(None, max_length=400)


class Objective(BaseModel):
    rank: int = Field(..., ge=1, le=20)
    name: str = Field(..., max_length=200)
    confidence: int = Field(..., ge=1, le=5)
    bullets: list[str] = Field(default_factory=list)
    beroe: str | None = Field(None, max_length=1200)
    sources: list[str] = Field(default_factory=list)


class DiscoveryQuestion(BaseModel):
    objective: str = Field(..., max_length=200)
    rank: int = Field(..., ge=1, le=20)
    person: str = Field(..., max_length=120)
    from_email: bool = False
    text: str = Field(..., max_length=600)
    # H46 — Category dropdown on Discovery questions. Free-form so prototype
    # categories (Commercial / Risk / People / Process / Sustainability /
    # Technology) and any future ones flow through without schema churn.
    category: str | None = Field(None, max_length=60)


class ValueAnchorPoint(BaseModel):
    text: str = Field(..., max_length=600)
    note: str | None = Field(None, max_length=400)


class ValueAnchor(BaseModel):
    objective: str = Field(..., max_length=200)
    points: list[ValueAnchorPoint] = Field(default_factory=list)


class EmailInsight(BaseModel):
    meta: str = Field(..., max_length=200)
    bullets: list[str] = Field(default_factory=list)


class PublicSignal(BaseModel):
    person: str | None = Field(None, max_length=120)
    headline: str = Field(..., max_length=240)
    text: str | None = Field(None, max_length=1200)
    url: str | None = Field(None, max_length=600)
    tag: str | None = Field(None, max_length=60)


class NewsItem(BaseModel):
    days_ago: int | None = Field(None, ge=0, le=3650)
    headline: str = Field(..., max_length=240)
    source: str | None = Field(None, max_length=120)
    signal: str | None = Field(None, max_length=600)
    url: str | None = Field(None, max_length=600)
    tag: str | None = Field(None, max_length=60)


class AnnualReportItem(BaseModel):
    title: str = Field(..., max_length=240)
    year: int | None = Field(None, ge=1900, le=2100)
    url: str | None = Field(None, max_length=600)
    bullets: list[str] = Field(default_factory=list)


class ClosingScenario(BaseModel):
    type: ScenarioType
    label: str | None = Field(None, max_length=80)
    text: str = Field(..., max_length=1200)


# ---------- Top-level read / write models ----------


class MeetingBriefOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    account_id: UUID

    call_type: BriefCallType | None = None
    call_date: date | None = None
    call_time: str | None = None
    call_platform: str | None = None
    call_duration_minutes: int | None = None

    win_condition: str | None = None
    cheat_sheet_win_condition_short: str | None = None

    company_snapshot: list[SnapshotStat] = Field(default_factory=list)
    call_timer: list[CallTimerSlot] = Field(default_factory=list)
    attendees: list[Attendee] = Field(default_factory=list)
    minefields: list[Minefield] = Field(default_factory=list)
    objectives: list[Objective] = Field(default_factory=list)
    discovery_questions: list[DiscoveryQuestion] = Field(default_factory=list)
    value_anchors: list[ValueAnchor] = Field(default_factory=list)
    email_insights: list[EmailInsight] = Field(default_factory=list)
    public_signals: list[PublicSignal] = Field(default_factory=list)
    news: list[NewsItem] = Field(default_factory=list)
    annual_reports: list[AnnualReportItem] = Field(default_factory=list)
    closing_scenarios: list[ClosingScenario] = Field(default_factory=list)
    cheat_sheet_never_say: list[str] = Field(default_factory=list)
    cheat_sheet_opening_asks: list[str] = Field(default_factory=list)
    # H46 — "Categories" tab in the prototype's brief. Free-form strings so
    # prototype vocab (commodity / category names) flows through.
    categories: list[str] = Field(default_factory=list)

    updated_at: datetime
    updated_by: UUID | None
    is_editable: bool = False


class MeetingBriefUpdate(BaseModel):
    """Whole-document PATCH — caller sends only the keys it wants to change."""

    call_type: BriefCallType | None = None
    call_date: date | None = None
    call_time: str | None = Field(None, max_length=120)
    call_platform: str | None = Field(None, max_length=120)
    call_duration_minutes: int | None = Field(None, ge=0, le=1440)

    win_condition: str | None = Field(None, max_length=1200)
    cheat_sheet_win_condition_short: str | None = Field(None, max_length=400)

    company_snapshot: list[SnapshotStat] | None = None
    call_timer: list[CallTimerSlot] | None = None
    attendees: list[Attendee] | None = None
    minefields: list[Minefield] | None = None
    objectives: list[Objective] | None = None
    discovery_questions: list[DiscoveryQuestion] | None = None
    value_anchors: list[ValueAnchor] | None = None
    email_insights: list[EmailInsight] | None = None
    public_signals: list[PublicSignal] | None = None
    news: list[NewsItem] | None = None
    annual_reports: list[AnnualReportItem] | None = None
    closing_scenarios: list[ClosingScenario] | None = None
    cheat_sheet_never_say: list[str] | None = None
    cheat_sheet_opening_asks: list[str] | None = None
    categories: list[str] | None = None
